hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/tensorflow/config/TensorBoardProcessMgr.java

```diff
@@ -0,0 +1,331 @@
+import java.math.BigInteger;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+/**
+ * *
+ * This class wraps a bash script with sudo rights that can be executed by the node['hopsworks']['user'].
+ * /srv/hops/domains/domain1/bin/tfserving.sh
+ * The bash script has several commands with parameters that can be executed.
+ * This class provides a Java interface for executing the commands.
+ */
+@Singleton
+@ConcurrencyManagement(ConcurrencyManagementType.CONTAINER)
+@DependsOn("Settings")
+public class TensorBoardProcessMgr {
+
+  private static final Logger logger = Logger.getLogger(TensorBoardProcessMgr.class.getName());
+
+  @EJB
+  private Settings settings;
+  @EJB
+  private HdfsUsersFacade hdfsUsersFacade;
+  @EJB
+  private HdfsLeDescriptorsFacade hdfsLeFacade;
+  @EJB
+  private HdfsUsersController hdfsUsersController;
+  @EJB
+  private TensorBoardFacade tensorBoardFacade;
+  @EJB
+  private ProjectFacade projectFacade;
+  @EJB
+  private UserFacade userFacade;
+
+  @PostConstruct
+  public void init() {
+  }
+
+  @PreDestroy
+  public void preDestroy() {
+
+  }
+  
+  @TransactionAttribute(TransactionAttributeType.NOT_SUPPORTED)
+  public TensorBoardDTO startTensorBoard(Project project, Users user, HdfsUsers hdfsUser, String hdfsLogdir)
+          throws IOException {
+
+    String prog = settings.getHopsworksDomainDir() + "/bin/tensorboard.sh";
+
+    Process process = null;
+    Integer port = 0;
+
+    BigInteger pid = null;
+
+    String tensorBoardDir = settings.getStagingDir() + Settings.TENSORBOARD_DIRS;
+
+    String localDir = DigestUtils.sha256Hex(project.getName() + "_" + hdfsUser.getName());
+    tensorBoardDir = tensorBoardDir + File.separator + localDir;
+
+    File tbDir = new File(tensorBoardDir);
+    if(tbDir.exists()) {
+      for(File file: tbDir.listFiles()) {
+        if(file.getName().endsWith(".pid")) {
+          String pidContents = com.google.common.io.Files.readFirstLine(file, Charset.defaultCharset());
+          try {
+            pid = BigInteger.valueOf(Long.parseLong(pidContents));
+            if (pid != null && ping(pid) == 0) {
+              killTensorBoard(pid);
+              try {
+                TensorBoard tb = tensorBoardFacade.findForProjectAndUser(project, user);
+                if(tb.getPid().equals(pid)) {
+                  tensorBoardFacade.remove(tb);
+                }
+              } catch(DatabaseException dbe) {
+                //skip
+              }
+            }
+          } catch(NumberFormatException nfe) {
+            logger.log(Level.WARNING, "Expected number in pidfile " +
+                    file.getAbsolutePath() + " got " + pidContents);
+          }
+        }
+      }
```

Comment:
Why do you have a "check and kill" if a TensorBoard is running here and another one in the TensorBoard.java class? Wouldn't be better for code and flow clarity to check only in a single place?

Reply:
TensorBoard.java is the entity class and does not have a check and kill. But I get your message, I've introduced a controller where I put will put the logic for cleanup.

Reply:
My bad, I meant `TensorBoardController.java` (line 77). You go through that code before going through this.

Reply:
Yes fixing that :+1: